rfd/0028-cluster-config-resources.md

```diff
@@ -0,0 +1,298 @@
+    // timeouts
+    int64 ClientIdleTimeout;
+
+    // DisconnectExpiredCert provides disconnect expired certificate setting -
+    // if true, connections with expired client certificates will get disconnected
+    bool DisconnectExpiredCert;
+
+    // KeepAliveInterval is the interval the server sends keep-alive messsages
+    // to the client at.
+    int64 KeepAliveInterval;
+
+    // KeepAliveCountMax is the number of keep-alive messages that can be missed
+    // before
+    // the server disconnects the connection to the client.
+    int64 KeepAliveCountMax;
+
+    // LocalAuth is true if local authentication is enabled.
+    bool LocalAuth;
+
+    // SessionControlTimeout is the session control lease expiry and defines
+    // the upper limit of how long a node may be out of contact with the auth
+    // server before it begins terminating controlled sessions.
+    int64 SessionControlTimeout;
+}
+```
+
+It is proposed to distribute them in the following fashion:
+
+```proto
+message ClusterMeta {
+    // This resource is not genuine configuration, it would contain only the
+    // common resource metadata.  The metadata name field would be used to
+    // represent the ClusterID.
+}
+
+message SessionRecordingConfig {
+    // ClusterConfig.SessionRecording is renamed to SessionRecordingConfig.Mode.
+    string Mode;
+    bool ProxyChecksHostKeys;
+}
+
+message ClusterNetworkingConfig {
+    int64 ClientIdleTimeout;
+    int64 KeepAliveInterval;
+    int64 KeepAliveCountMax;
+    int64 SessionControlTimeout;
+}
+
+// The already existing AuditConfig is turned into a standalone resource.
+message AuditConfig { ... }
+
+// DisconnectExpiredCert & LocalAuth fields are moved into ClusterAuthPreference.
+message ClusterAuthPreference {
+    ...
+    bool AllowLocalAuth;
+    bool DisconnectExpiredCert;
+}
+```
+
+#### Data migration & docs updates
+
+A new step should be added to `migrateLegacyResources` (called during auth
+server initialization) that takes care of migrating the `ClusterConfig` data
+stored in the backend to the new resource structures.  This probably implies
+that `ClusterConfig` as a structure needs to be kept around for a couple of
+subsequent Teleport versions for marshalling purposes.
+
+Documentation should be updated accordingly.
+
+### Restricting to a subset of values of a field
+
+Some users should not be capable of setting a field to any of the otherwise
+accepted values.  It is proposed to satisfy this requirement by introducing new
+RBAC verbs, interpreted as implying additional constraints. Such new verbs
+would be evaluated together with other RBAC verbs in the auth API handlers.
+
+In particular, this is needed for `SessionRecordingConfig` which has two
+fields:
+* `mode` with the possible values: `off`, `node`, `node-sync`,
+  `proxy`, `proxy-sync`;
+* a boolean `proxy_checks_host_keys`.
+
+Some users should be subject to the following restrictions:
```

Comment:
Well, one use case is precisely limiting the `session_recording` values & possibly also disabling the `second_factor` option, mentioned below.

Regarding `auth_connectors` RBAC, I don't think the situation is that simple. What seems to occur is that both connector-specific *and* `auth_connectors`-wide privileges need to be granted, so for a user to be able to update Github connectors they'd need something like:
```
  allow:
    rules:
    - resources: [github, auth_connector]
      verbs: [read, create, update]
```

Reply:
@alex-kovoy that's a valid concern and we considered new resources and verbs alternatives. We do need, however a fine-grained control over multiple independent parts of teleport, and having a monolith resource did not scale. So we are pursuing a strategy of splitting the resources at the extra cost of making it a bit harder to use, yes.

Reply:
@andrejtokarcik  I can confirm that `auth_connector` resource is a default way of restricting access to auth connectors. Not their individual parts.

@klizhentas splitting the resource is one thing but polluting the RBAC is another. Do we need it specifically for the cloud use cases? If so, it could be driven by the feature flags to prevent changes of specific config resources from API.

Reply:
I don't understand this concern with "RBAC polluting". We can provide more comprehensive resource-like subjects but allowing more fine-grained definitions seems much more in line with least privilege policies that we should strive to support.

Reply:
@andrejtokarcik my concern is that we would need to ship a book to teach people how to use RBAC and go around all types of resources that are there. This is why I would stick with a simple `cluster_config` RBAC subject and do the cloud specific constrains in the code for as long as it's manageable.